Generate observation tables

mapazarr opened the pull request:
The method still needs some work and clean-up. I am committing a preliminary version in order to trigger a discussion in the hangout tomorrow. Please don't look/comment on the minor details yet, but rather the API and the way of how to fill the obs table. I think maybe some other methods should be defined, like `FillOneObservation` and separate the building of a obs list from the faking of a list.
The method has a working test.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +70,90 @@ def sigma_energy_theta(energy, theta, sigma):
+    # random points between the start of 2010 and the end of 2014
+    #TODO: should this represent the time at the beginning of the run?
+    #      this has consequences for the ra/dec -> alt/az conversion
+    datestart = Time('2010-01-01T00:00:00', format='fits', scale='utc')
```
The test fails on travis-ci: https://travis-ci.org/gammapy/gammapy/jobs/65169219#L1197
You should see the same failure locally, no?

mapazarr replied:
@cdeil Sorry, the travis-ci took a bit longer than expected and then I forgot to check it.
The error
` ValueError: Format 'fits' is not one of the allowed formats ['astropy_time', 'byear', 'byear_str', 'cxcsec', 'datetime', 'decimalyear', 'gps', 'iso', 'isot', 'jd', 'jyear', 'jyear_str', 'mjd', 'plot_date', 'unix', 'yday']`
is due to a missing format in `~astropy.Time`:
http://astropy.readthedocs.org/en/latest/time/
Updating to the most recent version of astropy-dev solves the problem.
@cdeil : what is the rule about dependencies for astropy? Should we go with the latest version, or should we "freeze" a certain version from time to time?

cdeil replied:
@mapazarr On travis-ci we currently test with the latest dev and stable version of Astropy. The stable version is currently 1.0.2 (see https://travis-ci.org/gammapy/gammapy/jobs/65301895#L436). So you have to make this example work with Astropy 1.0.2.

In this case you can just use `format=iso` instead of `format=fits`, no?

mapazarr replied:
OK, I see. Yes I'll take iso for the moment.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +70,90 @@ def sigma_energy_theta(energy, theta, sigma):
     return psf
+
+
+def generate_observation_table(observatory, n_obs):
```
I would prefer `make_test_observation_table`.
For consistency I prefer `make` over `generate` (see [here](https://gammapy.readthedocs.org/en/latest/datasets/index.html#introduction)), and I'd put in the `test` because this is not for science analysis, but mainly for testing (and maybe for an example, because we don't have something better that's public.

cdeil replied:
You could add ~ a handful of parameters more to this function, e.g. the observatory name or the date range.

mapazarr replied:
Ok, I renamed it. I already have observatory for observatory name. As for other parameters, let's see if they are needed, and we can add them later.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +70,90 @@ def sigma_energy_theta(energy, theta, sigma):
+    # TODO: add column(s) for offset, and take it into account in the coord. transformation!!!
+
+    # RA, Dec
+    # random points on a sphere ref: http://mathworld.wolfram.com/SpherePointPicking.html
```
There's a utility function for this already: https://gammapy.readthedocs.org/en/latest/api/gammapy.utils.random.sample_sphere.html
Should be possible to re-use it here?

mapazarr replied:
Done.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +70,90 @@ def sigma_energy_theta(energy, theta, sigma):
+    # alt, az
+    # TODO: should they be in the ObservationTable? (they are derived quantities, like dead time)
+    # TODO: since I randomized RA/DEC without taking into account the observatory, I'm getting negative altitudes!!!
+    #       maybe I should randomize alt az, then transform to RA/DEC!!!
```
Drawing random observation positions in Alt / Az with Alt > 0 deg and then computing RA / DEC for those seems like a good idea to me. More realistic and not not much work.

mapazarr replied:
Done.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +70,90 @@ def sigma_energy_theta(energy, theta, sigma):
+
+
+    # TODO: operate row by row (using Observation class) instead of by columns?
+    # TODO: general methods for filling obs tables run by run; this function should call it.
```
Generally it should be preferred to have vectorized functions. If something can't be vectorised, then usually the right strategy is to put the Python for look as low as possible in the utility functions that are called.
But this can be tricky and sometimes it is the right decision to even have user-facing functions that are not vectorised.
Here I don't see any issue with using vectorised expressions everywhere?
We can talk about this tonight.

mapazarr replied:
I didn't understand what you mean.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +70,90 @@ def sigma_energy_theta(energy, theta, sigma):
+    """
+    n_obs_start = 1
+
+    astro_table = Table()
```
Is there a reason why you can't just instantiate an `ObservationTable` here and then fill it?
Does it do some checking in the constructor that prevents that pattern of creating it?
If no, I think I'd prefer that (a bit simpler).

mapazarr replied:
Fixed.

cdeil commented:
While you're working on this, could you pleas also update this observation table format spec / description?
https://gammapy.readthedocs.org/en/latest/dataformats/observation_lists.html

Don't worry too much about what's in `runinfo.fits` now ... the important thing is to come up with a good, documented format. Once that is written down we can adapt the existing file we have to this format.

cdeil commented:
It would be nice if you could fake TELESCOPE_PATTERN (or at least N_TEL) and MUON_EFFICIENCY as well, these are probably parameters we want to use to group observations soon.

Otherwise :+1: , this is heading in the direction I had in mind when we talked about this.

mapazarr commented:
I fixed the error on travis-ci. I will continue working on this tomorrow.

mapazarr commented:
@cdeil @mimayer
Do we need TSTOP and ONTIME? We can calculate ONTIME as TSTOP - TSTART, right?

cdeil commented:
I'd say make `TSTART` required. You want to know the run date e.g. for time selection in findruns "give me all runs from 2005 to 2007".
`TSTOP` and `ONTIME` are sometimes of interest, so should be optional. You can still define them in this format, so that everyone will use the same names if they do add those columns, and not one guy calls it `TSTOP` and then another gal calls it `TEND`.

mapazarr commented on gammapy/utils/tests/test_time.py:
```diff
@@ -0,0 +1,21 @@
+from gammapy.utils.time import time_ref_from_dict #OK!
+#from . import time_ref_from_dict
+#from .. import time_ref_from_dict
+
```
@cdeil which relative import should I use here?

cdeil replied:
Das sollte gehen:
```python
from ..time import time_ref_from_dict
```

mapazarr replied:
Ok fixed. Thanks!

mapazarr commented on gammapy/utils/tests/test_time.py:
```diff
@@ -0,0 +1,21 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
```
Do I need the "future" here? :-)

cdeil replied:
Yes, please add this to any new file.

It's an Astropy coding guideline, and it also makes sense: if you don't add it, it is possible to get slightly different behaviour on Python 2 and 3 (e.g. `1/2 ==0` on Python 2 or `1/2 == 0.5` on Python 3). If test coverage is not very high and you don't put in the right test dataset, issues can go unnoticed.

I know it's annoying boilerplate code, but thinking about / discussing where you need or might need it in the future is worse than just pasting it into every file.

cdeil replied:
We can remove it when we drop Python 2 support. I.e. when Sherpa supports Python 3 and > 90% of our users are on Python 3, i.e. maybe in a few years, maybe never. :-)

mapazarr replied:
Ok.

mapazarr commented:
@cdeil I started working on the doc about the format for obs lists. Please have a look at the "TODOs" in the code.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -22,23 +22,48 @@ This has the advantage that everyone can work with whatever tool they like. Here
-   1234,24,2013-03-22 14:32,1832,83.7,-5.2,Crab Nebula,32
+Usually it has many more columns with information about each observation. A list of Gammapy supported columns is::
+
+   OBS_ID: observation ID as an integer (starting at 0 or 1? does it matter?)
```
Can you please re-format this as an RST simple table?
See [here](http://sphinx-doc.org/rest.html#tables) or [here](http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#simple-tables) for a description of the format and [here](http://astropy.readthedocs.org/en/latest/time/index.html) for an example of usage in the Astropy docs. Note that on any Astropy Sphinx docs page you can click "Page Source" at the very bottom and see the RST file ... this is sometimes easier to learn how to do something than to read the RST or Sphinx docs.

cdeil replied:
`(starting at 0 or 1? does it matter?)` -> every experiment / observatory will do this as they like, no need to explain / define anything how OBS_ID is filled here, I think.

cdeil replied:
The table should also have a `Unit` column, because in CSV there's no way to store the unit and something has to be defined.

mapazarr replied:
Done, done, done.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -22,23 +22,48 @@ This has the advantage that everyone can work with whatever tool they like. Here
 * Your own script using e.g the `Python standard library CSV module <http://docs.python.org/2/library/csv.html>`_ or `pandas <http://pandas.pydata.org>`_
 
-A run list must have at least a column called ``Run``::
+A run list must have at least a column called ``OBS_ID``::
```
Can you re-structure the whole page a bit?

Gammapy can work with observation tables in whatever format Astropy Table can read. So it could be CSV, but that doesn't have meta info, so at least for data distribution we'll use FITS (the current `runinfo.fits`).

So this general info about CSV tools should go to https://gammapy.readthedocs.org/en/latest/dataformats/file_formats.html#csv (if it's not all there already) and if you think it's useful, just keep a Sphinx `:ref:` link to that section from here.

mapazarr replied:
Done: (re)moved file format stuff.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -22,23 +22,48 @@ This has the advantage that everyone can work with whatever tool they like. Here
+   OBS_ID: observation ID as an integer (starting at 0 or 1? does it matter?)
+   RA: pointing position right ascension in equatorial (ICRS) coordinates
+   DEC: pointing position declination in equatorial (ICRS) coordinates
+   AZ: average azimuth angle during the observarion
```
typo: `observarion` -> `observation` (same in next line)

mapazarr replied:
Fixed.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -22,23 +22,48 @@ This has the advantage that everyone can work with whatever tool they like. Here
+   AZ: average azimuth angle during the observarion
+   ALT: average altitude angle during the observarion
+   MUON_EFFICIENCY: average muon efficiency of the telescopes
+   TIME_START: start time of the observation stored as number of seconds after the time reference in the header
```
I don't know if RST tables can nicely accomodate such long descriptions. If they automatically wrap to multiple rows or there's a way to do that, great. If no, please put a short description here and then a bullet list with longer explanations on some things below the table.

mapazarr replied:
Yes, they can.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -22,23 +22,48 @@ This has the advantage that everyone can work with whatever tool they like. Here
+   TIME_STOP: end time of the observation in the same forma as TIME_START
+   TIME_OBSERVATION: duration of the observation
+   TIME_LIVE: duration of the observation without dead time
+   TRIGGER_RATE: average trigger rate of the system
```
Things like trigger rate, temperature, quality should be optional.
Please add a column with flag "required" or "optional" to the table.
No idea for a good name for this column at the moment.

mapazarr replied:
I find it difficult to label the columns as required or optional.
Actually, we already said before that the only required column is `OBS_ID`.
I guess it depends on what do you want to do with the table afterwards. If you want to run an analysis, the only column needed is the `OBS_ID`: gammapy should be able to find the corresponding observation lists.
If you want to make a selection of runs (apply some filter, like findruns does), you need the columns that you want to filter on.
Plus, chances are that we label a column as optional, then implement some tool that works on it and forget to update the doc.

cdeil replied:
This optional / required question is related to the question whether we have one or two classes representing observation tables / data store index tables. I still don't know what is best. If there's only one class, I agree that only `OBS_ID` should be required.

My suggestion would be you continue implementing functionality (obs selection and grouping) for a few days and then maybe it will become clearer what is best?

mapazarr replied:
Ok.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -22,23 +22,48 @@ This has the advantage that everyone can work with whatever tool they like. Here
+
+Extra user defined columns are allowed; Gammapy will ignore them.
+
+In order for the extra columns to have full meaning a header is needed with at least the following keywords::
```
Again, please make the header keywords a table.

mapazarr replied:
done

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -22,23 +22,48 @@ This has the advantage that everyone can work with whatever tool they like. Here
+   TIME_REF_MJD_INT: reference time for other times in the list (i.e. TIME_START/TIME_STOP). Integer value in mean julian days.
+   TIME_REF_MJD_FRA: fraction of integer value defined in TIME_REF_MJD_INT.
+
+TODO: change names already defined in event lists (and in `~gammapy.utils.time.time_ref_from_dict`) ``MJDREFI`` and ``MJDREFF``? Or adopt ``TIME_REF_MJD_INT`` and ``TIME_REF_MJD_FRA``?
```
Yes, I think `MJDREFI` and `MJDREFF` is an existing standard or convention. We should re-use it here.

mapazarr replied:
Done

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -22,23 +22,48 @@ This has the advantage that everyone can work with whatever tool they like. Here
+
+TODO: change names already defined in event lists (and in `~gammapy.utils.time.time_ref_from_dict`) ``MJDREFI`` and ``MJDREFF``? Or adopt ``TIME_REF_MJD_INT`` and ``TIME_REF_MJD_FRA``?
+
+TODO: should the observation list have already a header? Or should the header values be defined as columns, then interpreted correspondingly when stored into an ObservationTable? (I might be mixing up 2 things: obs lists and obs tables here? or should they have the same format?)
```
I'm not sure what you mean here.

When ObservationTable is read from FITS it will have the FITS header in the `.meta` dict and this should be used.
When ObservationTable is read from CSV there will be no header info.
We should not try to store header key / vals in columns.

I realise there's very different things. A "run list" can be just a list of `OBS_ID`s. A data store index table can contain lots of information that will be needed for `findruns`. As I said, I don't have a full concept yet how best to implement classes that cover all the common use cases of things people do with observation lists for Gammapy.
Here my suggestion would be that you try to implement the common use cases as examples in the tests and see how it goes. Probably this will give clues as to which classes need to be added and if any should be removed.

mapazarr replied:
Ok. We can discuss tomorrow in the hangout how to proceed.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -22,23 +22,48 @@ This has the advantage that everyone can work with whatever tool they like. Here
+
+TODO: should the observation list have already a header? Or should the header values be defined as columns, then interpreted correspondingly when stored into an ObservationTable? (I might be mixing up 2 things: obs lists and obs tables here? or should they have the same format?)
+
+TODO: should the observation list already define the times in this ``TIME_REF_MJD_INT + TIME_REF_MJD_FRA`` format? Or should it be converted once its read into an ObservationTable?
```
I'm not sure in what format we should store times in observation tables.
* MET (mission elapsed time) is a simple float, but needs the reference time to be meaningful, and it's not possible to store that in CSV files, so I guess is not a great choice.
* MJD (modified Julian date) is a simple float and doesn't need the reference time, but maybe is not very readable.
* UTC time string is longer, but maybe most readable and what user's want?

I'd say, maybe use UTC time strings (i.e. recommend some format, but accept any string the `astropy.time.Time` will take as input?

mapazarr replied:
Well, the variables are called "MJD" (`MJDREFI`, `MJDREFF`), so we can't just put UTC on top of it.
My table generator uses MJD as default, but has a debug option to show UTC.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -22,23 +22,48 @@ This has the advantage that everyone can work with whatever tool they like. Here
+
+TODO: should the observation list already define the times in this ``TIME_REF_MJD_INT + TIME_REF_MJD_FRA`` format? Or should it be converted once its read into an ObservationTable?
+
+TODO: ``TEL_LIST``: incompatibility wit CSV format, since it's a CSV list of tels???!!!
```
You can have comma-separated values in a column in a CSV file. Quoting takes care of that.
But this is not very user-friedly to edit in a CSV file.
The HESS integer telescope pattern is also not very user-friendly and I don't know if it even scales to the 100 CTA telescopes. A bitpattern `100010110101` would be another option, less human-readable than the comma-separated telescope numbers, but better to store in a CSV file.

I'd say just choose something here. And please file a new issue that in the future we should add a class to convert between various formats to store TELARRAY specifications and then re-think how to store this in observation lists. Note that the HAP-exported event list files have a TELARRAY extension (see https://gammapy.readthedocs.org/en/latest/api/gammapy.data.TelescopeArray.html), you could also have a look at that and implement that functionality on that class. But maybe best to defer to a future issue / pull request?

mapazarr replied:
Done: issue #282 has been created, and a reference put into the doc.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -22,23 +22,48 @@ This has the advantage that everyone can work with whatever tool they like. Here
+TODO: ``TEL_LIST``: incompatibility wit CSV format, since it's a CSV list of tels???!!!
 
-Special column names that the Gammapy analysis tools understand:
+TODO: how to biuld sphinx only for a specific module or file?
```
I'm now aware of a way to build the Sphinx docs for just a part of the files. If you figure out a way to do that, please let me know. It's annoyingly slow.

mapazarr replied:
I agree.
The sphinx doc says it should be possible right before this section:
 http://sphinx-doc.org/invocation.html#makefile-options
But I haven't been able to get it to run through our `setup.py` file.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +71,121 @@ def sigma_energy_theta(energy, theta, sigma):
+    Parameters
+    ----------
+    observatory : string
+    	name of the observatory
```
You have to somehow tell the user what the choices are.
This is complicated in this case because `observatory_locations` doesn't show up in the API docs (see issue #111), so you can't refer to there from here.
So maybe all we can do for now is leave a TODO in the docstring and mention `observatory_locations` without linking to it?

mapazarr replied:
I added a reference, without leaving a TODO, because chances are that at some point the list will appear in the doc, and we forget to remove the TODO.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +71,121 @@ def sigma_energy_theta(energy, theta, sigma):
+    #       calculate pointing observation (ra dec), considering the observatory location, and alt az is at the middle of the observation
+    #       then convert to alt az, considering the observatory location
+
+    # TODO: is there a way to comment on the column names?!!!!!
```
@mapazarr – The `astropy.table.Column` class has a description, and you can set it here.
The problem is that CSV and even FITS don't support it:
```
from astropy.table import Table, Column
t = Table()
c = Column(name='aaa', data=[1,2,3], unit='m/s', description='Triple a')
t.add_column(c)
t.write('/tmp/aaa.fits')
from astropy.io import fits
header = fits.getheader('/tmp/aaa.fits')
header = fits.getheader('/tmp/aaa.fits', ext=1)
print(repr(header))
```
shows that units are saved in FITS, but description is not:
```
TTYPE1  = 'aaa     '                                                            
TFORM1  = 'K       '                                                            
TUNIT1  = 'm s-1   '                                                            
```
Now sometimes FITS tables do contain description comments of header keywords and columns, but with Astropy table these are lost when you do:
```
Table.read('file.fits').write('file2.fits')
```
I don't know if / how this will improve in the future, for now I'd say it's OK if you don't add a column description here, but it could also be useful in case the Astropy Table / FITS integration improves in the future or if we decide to use a format that can store all the Astropy Table meta-data like [ECSV](http://astropy.readthedocs.org/en/latest/api/astropy.io.ascii.Ecsv.html).

mapazarr replied:
Ok, I was thinking in the case the names of the columns were very cryptic. So I guess it's fine if we don't add them for now.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +71,121 @@ def sigma_energy_theta(energy, theta, sigma):
+    # TODO: store obs name as a column or as a header value?!!!
+    # TODO: format times!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+    # TODO: alt az at mean time!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+    # TODO: restrict to night time? (or dark time?)!!!!!!!!!!!!
```
I'd say don't bother making these fake tables more realistic, leave as is and move on to the observation planning toolbox. Once `astroplan` is in place I plan to do https://github.com/gammapy/gammapy/issues/79#issue-27908311 as a use case and then it will get much more realistic. For now the purpose is only to test the observation handling toolbox you implement, and for this you don't need a good fake.

mapazarr replied:
Ups! :-) too late! I already did.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +71,121 @@ def sigma_energy_theta(energy, theta, sigma):
+    sky_coord = alt_az_coord.transform_to(FK5)
+    ra = sky_coord.ra
+    col_ra = Column(name='RA', data=ra)
+    obs_table.add_column(col_ra)
```
I think some of these lines can be combined to get a shorter yet not less readable function:
```python
obs_table.add_column(Column(name='RA', data=skycoord.ra))
```

mapazarr replied:
done

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +71,121 @@ def sigma_energy_theta(energy, theta, sigma):
+
+
+def make_test_observation_table(observatory, n_obs):
+    """Generate an observation table.
```
```python
"""Make a test observation table.
...
"""
```

mapazarr replied:
Done.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -43,6 +43,12 @@ class ObservationTable(Table):
+    * ``AZ``
+    * ``ALT``
+    * ``RA``
+    * ``DEC``
```
Link to the format description in the docs (a Sphinx `:ref:` directive should work) instead of trying to duplicate (part of) the information here in this docstring.

mapazarr replied:
I thought so. I wasn't sure however how to make the proper link. Thanks for the tip.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -56,6 +62,7 @@ def info(self):
         ss += 'Total live time: {}\n'.format(livetime)
         dtf = 100. * (1 - livetime / ontime)
         ss += 'Average dead time fraction: {:5.2f}%'.format(dtf)
+        #TODO: units are not shown!!!
```
Agreed, units should be shown. Please fix this.

mapazarr replied:
Fixed.

cdeil commented on gammapy/utils/random.py:
```diff
@@ -30,6 +30,14 @@ def sample_sphere(size, lon_range=None, lat_range=None, unit='radians'):
         Longitude and latitude coordinate arrays
     """
+    # Correctly interpret units at an early stage
+    if unit in ['rad', 'radian', 'radians']:
```
We should use the `astropy.coordinates.Longitude` and `astropy.coordinates.Latitude` classes, which take care of unit handling. Can you please rewrite this function (in a separate PR or here, as you like) to use the Astropy classes instead? 

There's probably a few callers to this function which you can find e.g. via the full-text search on Github. Please fix these also.

PS: quite a bit of code in Gammapy was written before `Quantity` existed, i.e. needs to be updated to use quantities where appropriate.

mapazarr replied:
OK. Maybe on a separate pull request.
Issue #283. (for some reason I can't assign it to myself?)
Can you remind me how to make a full-text search on Github?

cdeil replied:
@mapazarr – Only people with write access to repos can add labels and assign issues on Github. For full-text search, just go to https://github.com/gammapy/gammapy and type in your question in the search field. Here's an example: https://github.com/gammapy/gammapy/search?utf8=%E2%9C%93&q=observation

The Github search is awesome, but there are cases where it doesn't do what you naively expect. In those cases, or when you're unsure how to best perform a search, head over here: https://help.github.com/categories/search/

mapazarr replied:
Thanks for the info.

cdeil commented on gammapy/utils/tests/test_time.py:
```diff
@@ -0,0 +1,14 @@
+    time_ref_dict = dict(MJDREFI=mjd_int, MJDREFF=mjd_frac)
+    time_ref = time_ref_from_dict(time_ref_dict)
+    decimal = 4
+    s_error = "time reference not compatible with defined values"
```
Please don't add error messages to tests. It adds a cost (work to write / review them and read them when you read a test) and almost no benefit (tests that pass once very rarely fail, and if they do the traceback and looking at the code is what's useful, not some printed message).

mapazarr replied:
Fixed.

cdeil commented on gammapy/utils/tests/test_time.py:
```diff
@@ -0,0 +1,14 @@
+    time_ref = time_ref_from_dict(time_ref_dict)
+    decimal = 4
+    s_error = "time reference not compatible with defined values"
+    np.testing.assert_almost_equal(time_ref.mjd, mjd_int + mjd_frac, decimal, s_error)
```
Pleas use the import pattern we use everywhere in Gammapy:
```python
from numpy.testing import assert_almost_equal

assert_almost_equal(...)
```

mapazarr replied:
Fixed.

cdeil commented on gammapy/utils/time.py:
```diff
@@ -0,0 +1,30 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""Time related utility functions."""
```
Please add the dummy `from __future__ import ...` line, even if it doesn't matter here at the moment.

mapazarr replied:
Fixed.

cdeil commented on gammapy/utils/time.py:
```diff
@@ -0,0 +1,30 @@
+
+
+def time_ref_from_dict(meta):
+    """ Calculte the time reference from metadata.
```
No space after `"""`.
Typo: `Calculte` -> `Calculate`

mapazarr replied:
Fixed.

cdeil commented on gammapy/utils/time.py:
```diff
@@ -0,0 +1,30 @@
+    Parameters
+    ----------
+    meta : dict
+    	dictionnary with the keywords 'MJDREFI' and 'MJDREFF'
```
Typo: `dictionnary` -> `Dictionary`

mapazarr replied:
fixed

cdeil commented on gammapy/utils/time.py:
```diff
@@ -0,0 +1,30 @@
+
+    Returns
+    -------
+    `~astropy.Time` object with the reference time in units of MJD.
```
Does the formatting look OK here in the html docs output?

The Astropy time lingo is "format" for MJD. Please use that here instead of "in units of MJD".

mapazarr replied:
Fixed.

cdeil commented:
@mapazarr – I've left a lot of inline comments. I hope they are helpful to you to continue with this.

There's two options:
- Try to finish up what's here (format docs page & make test table function) on Monday or Tuesday.
- Continue working on this PR for the next week, and start implementing the functionality we talked about.

Each approach has pros and cons ... up to you.

mapazarr commented:
Ok, I did a major revision. Please have a look. I'd like to have the PR merged soon, since this list is getting too long an a bit unclear (unuebersichtlich).

mapazarr commented:
I noticed the problem with the pull request, but I'm trying a rebase and I get the status `Everything up-to-date`, so nothing changes. I tried to emulate the procedure on PR #275 without success. What else can I try? There is also no information about what are the merge conflicts...

cdeil commented:
@mapazarr – Thanks for your work on this. I agree it's best to get this merged soon and then you can continue in other pull requests. 

One general comment: the commit history here is messy (see https://github.com/gammapy/gammapy/pull/278/commits), there's a lot of "clean up" and "I forgot" commits. I don't care so much about a clean commit history for Gammapy, so I'm willing to merge this as-is. But Astropy and most other projects do care about their commit history, it especially gets important when a project has bugfix releases (like Astropy 1.0.4, we don't do bugfix releases in Gammapy at the moment) and a maintainer has to go through the commit log and decide which commits to backport to the stable branches.

There's a ton of articles how to achieve clean commit logs with git. I usually use `git rebase --interactive master` when a pull request is close to final and then squash, reorder, rename commits as needed. But this does require that from the commit message it's clear what's in each commit, and that is not the case with "renamed a few things". This is a good read: https://github.com/blog/2019-how-to-undo-almost-anything-with-git

If you do want to attempt to learn more about git by cleaning up the commits here, I suggest you make a copy of this branch or your whole repo on your disk before, because there's the opportunity to screw up with git rebase and loose valuable edits and commits. So as I said, my suggestion here would be to leave it alone and start doing this in future PRs.

cdeil commented on docs/dataformats/file_formats.rst:
```diff
@@ -194,6 +194,9 @@ XML
 **XML** files (see `Wikipedia <http://en.wikipedia.org/wiki/Xml>`__)
 
+GammaLib / ctools uses an "observation definition" XML format described
+`here <http://gammalib.sourceforge.net/user_manual/modules/obs.html#describing-observations-using-xml>`__.
```
Please point here http://cta.irap.omp.eu/gammalib-devel/ instead of to sourceforge.

cdeil replied:
This comment hasn't been addressed, please point to http://cta.irap.omp.eu/gammalib-devel/  instead.

(What I usually do before asking for final review and merge on a PR is to do a final review on Github myself, i.e. look at the `Conversation`, `Commits` and `Files changed` tabs. Github only shows inline comments where the corresponding line hasn't been updated by a later commit. Sometimes this means that one comment hasn't been addressed, sometimes the conclusion of the discussion was that no change is needed.

mapazarr replied:
Sorry. There were so many that I missed it. I also look everywhere, when the PR is ready for the final review.
Done.

cdeil commented:
@mapazarr - For PRs where you edited docs a lot (either RST pages and / or docstrings), like this one, please put the html docs online and add a link to them from the pull request.
This makes the PR easier to review.

In this case I did fetch your branch locally and did `python setup.py build_sphinx` and then copy the `docs/_build/html` folder to Dropbox. This is the first time I use Dropbox to host a static webpage, @mapazarr, can you see it here?
https://dl.dropboxusercontent.com/u/4923986/temp/gammapy-278/html/dataformats/observation_lists.html

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -10,38 +10,88 @@ the information in the run list. These filenames can either be given directly
+
+Example
+-------
+The tool `~gammapy.datasets.make.make_test_observation_table` can generate a `~gammapy.obs.ObservationTable` with dummy values.
```
For Sphinx links to API entries to work, you have to point to the "public" location (see [here](https://gammapy.readthedocs.org/en/latest/development/index.html#where-should-i-import-from) and [here](https://dl.dropboxusercontent.com/u/4923986/temp/gammapy-278/html/api/gammapy.datasets.make_test_observation_table.html).
In this case: `~gammapy.datasets.make_test_observation_table`

mapazarr replied:
Fixed.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +72,158 @@ def sigma_energy_theta(energy, theta, sigma):
+    col_obs_id = Column(name='OBS_ID', data=obs_id)
+    obs_table.add_column(col_obs_id)
+
+    # TODO: `~astropy.table.Table` doesn't handle `~astropy.time.TimeDelta` columns
```
I don't see any reason why we should use `TimeDelta`. Why not simply use `Quantity` and remove this TODO?

mapazarr replied:
Done.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +72,158 @@ def sigma_energy_theta(energy, theta, sigma):
+
+    # obs time: 30 min
+    time_observation = TimeDelta(30.*60.*np.ones_like(col_obs_id.data), format='sec')
+    time_observation = Quantity(time_observation.sec, 'second') # converting to quantity
```
It's simpler if you just use Quantity:
```
 Quantity(30, 'min').to('second')
```

mapazarr replied:
Done.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +72,158 @@ def sigma_energy_theta(energy, theta, sigma):
+    # start of the night, when generating random night hours
+    datestart = Time('2010-01-01 00:00:00', format='iso', scale='utc')
+    dateend = Time('2015-01-01 00:00:00', format='iso', scale='utc')
+    time_start = Time((dateend.mjd - datestart.mjd)*np.random.random(len(col_obs_id)) + datestart.mjd, format='mjd', scale='utc')
```
Please roughly respect the 80 character line length limit.
Use temp variables to break up complex expressions in pieces.

Running pep8 on the files you've edited will show you this as a PEP8 violation.
Good Python editors / IDEs like PyCharm will also directly indicate this to you in the editor.

mapazarr replied:
Fixed.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +72,158 @@ def sigma_energy_theta(energy, theta, sigma):
+    time_start = Time(time_start_i, format='mjd', scale='utc')
+
+    # random generation of night hours: 6 h (from 22 h to 4 h), leaving 1/2 h time for the last run to finish
+    night_start = TimeDelta(22.*60.*60., format='sec')
```
The whole point of Quantity and Time is that you shouldn't have to manually do `*60.*60.`.
Please use Quantity.

mapazarr replied:
Done.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +72,158 @@ def sigma_energy_theta(energy, theta, sigma):
+    # optional: make it depend on other pars: temperature, pressure, humidity,...
+    sky_coord = alt_az_coord.transform_to(FK5)
+    ra = sky_coord.ra
+    col_ra = Column(name='RA', data=ra)
```
Please use this pattern more:
```python
table['RA'] = sky_coord.ra
```

mapazarr replied:
Done.

cdeil commented on gammapy/utils/time.py:
```diff
@@ -0,0 +1,55 @@
+
+
+def time_ref_from_dict(meta):
+    """Calculate the time reference from metadata.
```
Always add empty line after the first one-line description.

mapazarr replied:
Done.

cdeil commented on gammapy/utils/time.py:
```diff
@@ -0,0 +1,55 @@
+
+    Returns
+    -------
+    `~astropy.time.Time` object with the reference time in format of `MJD`.
```
This doesn't look too bad in the html docs, see [here](https://dl.dropboxusercontent.com/u/4923986/temp/gammapy-278/html/api/gammapy.utils.time.time_ref_from_dict.html).

But I still think the normal pattern is to have the same format for "Returns" as for "Parameters", in this case:
```
time : `~astropy.time.Time`
    Reference time with ``format='MJD'``.
```
Or did you see this other format for "Returns" docstrings in many places in Numpy / Scipy / Astropy?

mapazarr replied:
Ok.

cdeil commented:
@mapazarr – I've left a few comments, mostly concerning trivial stuff.

The big question we should discuss in the hangout is whether we want to stick with the time format which requires a header: https://dl.dropboxusercontent.com/u/4923986/temp/gammapy-278/html/dataformats/observation_lists.html
The advantage is that it's identical to what's done in event lists.

The big drawback is that it means observation tables can only be saved and exchanged in formats that have headers, i.e. not in CSV.

The alternative would be to put a time UTC string. Performance is not an issue here and I think I'd prefer this, but I'm not 100% sure. @mapazarr - What do you think?

The other header key, observatory, could also be stored as a column if needed. That's OK, because storage space and I/O speed are irrelevant for observation lists. This might even be advantageous, because it opens up the possibility to have a combined HESS / VERITAS observation list, or probably more useful, to store information which HESS chain / cuts should be used when one wants, i.e. add all the info needed to be able to run analyses via the `DataStore` class.

Let's discuss tonight and finalise this PR, but keep the possibility open to change formats and contents in the coming months as we see opportunities for improvement.

mapazarr commented:
OK, I implemented the last few comments. Could you please review and merge the PR?

For the doc: google drive is not very efficient: takes a lot of time to upload and needs to make public every single file, plus the links within the files are not respected. I have created a new dir in google drive called `gammapy_html_doc` and sent an invitation to @cdeil and @adonath. I put a tar ball with the generated doc. Please check if you can access it and if it is readable.

I also rebased.

cdeil commented:
@mapazarr – IMO a tarball with the docs is not helpful for PR review. It takes a few minutes for me to get the latest version of your branch with git, run `python setup.py build_sphinx` and review the docs locally.

The two advantages of having the docs for a PR online would be:
- Might get more people to have a look at your work and give feedback. Can link from Github comments to sections in the HTML docs to explain suggested improvements.
- Saves the reviewer a few minutes to get at the HTML docs.

So I'm OK if you don't post HTML versions of the docs for future PRs, it was just a suggestion.
If you decide you want to do it, please use Github pages or make a Dropbox account or use some other place where you can put static HTML pages for free.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -10,38 +10,87 @@ the information in the run list. These filenames can either be given directly
+AZ                average azimuth angle during the observation                                                      yes?
+ALT               average altitude angle during the observation                                                     yes?
+MUON_EFFICIENCY   average muon efficiency of the telescopes                                                         yes?
+TIME_START        start time of the observation stored as number of seconds after the time reference in the header  yes?
```
Instead of writing "number of seconds after the time reference in the header" you should say MET.

And ideally `MET` would link to a "Glossary" section or a "Time in Gammapy" section.

@mapazarr – please add this and write a few sentences (linking to the Fermi description of time as done [here](https://hess-confluence.desy.de/confluence/display/HESS/HESS+Open+Source+Tools+-+Data+Formats+and+Exporters#HESSOpenSourceTools-DataFormatsandExporters-TimeinHESSdataanalysis)).
A "Time in Gammapy" section could go here: https://gammapy.readthedocs.org/en/latest/dataformats/index.html
A "Glossary" (for now only with one entry "MET") could go here: https://gammapy.readthedocs.org/en/latest/references.html
(make a sub-title "Papers" for what is there now and add the "Glossary" below).

I think this is not much work and doesn't need much discussion, i.e. you could add this in this PR and it would still be merged within ~ a day.

mapazarr replied:
Ok, done.
Does this mean we should use the same time reference as Fermi?
http://fermi.gsfc.nasa.gov/ssc/data/analysis/documentation/Cicerone/Cicerone_Data/Time_in_ScienceTools.html
ref: midnight at the beginning of January 1, 2001, in the UTC system
I think that "mission elapsed time" (MET) is a bit confusing, since I would imagine each experiment would have a different reference, depending on the starting date. Or does this refer to a well established system, where everyone uses the above defined time reference?

cdeil replied:
@mapazarr – Every mission can choose their own reference time point where `MET=0`.
See https://heasarc.gsfc.nasa.gov/cgi-bin/Tools/xTime/xTime.pl for what different missions chose.

For HESS Karl, Michael and I (as the people working on FITS exporters) made the choice to use the same as Fermi LAT. It's not unlikely CTA (or MAGIC, VERITAS, HAWC) will choose a different one.

But the point is that the data format and software processing should be independent of this choice. This is achieved by putting the reference time point in the file header and then using MET values.

Makes sense?

mapazarr replied:
OK, that's what I thought. So we don't make any suggestions about the origin in our doc. I was confused because of the reference to the Fermi website. Now I see, the website is meant only for the MET description, but not for the MET origin. I could add a sentence: each experiment will have a different MET origin that should be included in the header of the data files. What do you think?

cdeil replied:
Yes. Just "will have" -> "can have".

mapazarr replied:
Done.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -10,38 +10,87 @@ the information in the run list. These filenames can either be given directly
+ALT               average altitude angle during the observation                                                     yes?
+MUON_EFFICIENCY   average muon efficiency of the telescopes                                                         yes?
+TIME_START        start time of the observation stored as number of seconds after the time reference in the header  yes?
+TIME_STOP         end time of the observation in the same forma as TIME_START                                       yes?
```
Typo: `forma` -> `format`

mapazarr replied:
Fixed.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -10,38 +10,87 @@ the information in the run list. These filenames can either be given directly
+TRIGGER_RATE      average trigger rate of the system                                                                no
+MEAN_TEMPERATURE  average temperature of the atmosphere                                                             no
+N_TELS            number of telescopes participating in the observation                                             yes?
+TEL_LIST          string with a CSV list of telescope IDs participating in the observation                          yes?
```
"CSV" -> "comma-separated"

mapazarr replied:
Added to the glossary, otherwise the line starts getting too long.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -10,38 +10,87 @@ the information in the run list. These filenames can either be given directly
+MEAN_TEMPERATURE  average temperature of the atmosphere                                                             no
+N_TELS            number of telescopes participating in the observation                                             yes?
+TEL_LIST          string with a CSV list of telescope IDs participating in the observation                          yes?
+QUALITY           data quality; recommended: "spectral" or "detection" (not used by Gammapy at the moment)          no
```
Remove "(not used by Gammapy at the moment)". There's more things like `TRIGGER_RATE`, ... that aren't used, no need to try and convey that info here.

mapazarr replied:
Done.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -10,38 +10,87 @@ the information in the run list. These filenames can either be given directly
-* `TOPCAT <http://www.star.bris.ac.uk/~mbt/topcat/>`_ or `STILTS <http://www.star.bris.ac.uk/~mbt/stilts/>`_
-* `csvkit <https://csvkit.readthedocs.org/en/latest/>`_
-* Your own script using e.g the `Python standard library CSV module <http://docs.python.org/2/library/csv.html>`_ or `pandas <http://pandas.pydata.org>`_
+Extra user defined columns are allowed; Gammapy will ignore them.
```
"user defined" -> "user-defined"

mapazarr replied:
Done.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -10,38 +10,87 @@ the information in the run list. These filenames can either be given directly
+keyword           description
+================  ==========================================================================================================================================================================================================
+OBSERVATORY_NAME  name of the observatory where the observations were taken. This is important for instance for coordinate transformations between celestial (i.e. RA/dec) and terrestrial (i.e. az/alt) coordinate systems.
+MJDREFI           reference time for other times in the list (i.e. TIME_START/TIME_STOP). Integer value in mean julian days.
```
For `MET` and the reference time it really would be best to have a single place where this is explained in Gammapy.
Then link to it from here and from the event list format specification and class, instead of trying to quickly re-explain this in several places.

mapazarr replied:
Ok, added a reference to "Time in Gammapy" section here:
 https://gammapy.readthedocs.org/en/latest/dataformats/index.html

cdeil replied:
Thanks for doing these extra improvements in this PR!

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -10,38 +10,87 @@ the information in the run list. These filenames can either be given directly
+================  ==========================================================================================================================================================================================================
+keyword           description
+================  ==========================================================================================================================================================================================================
+OBSERVATORY_NAME  name of the observatory where the observations were taken. This is important for instance for coordinate transformations between celestial (i.e. RA/dec) and terrestrial (i.e. az/alt) coordinate systems.
```
I would suggest to add a `Required?` column here as well.

`OBSERVATORY_NAME` is optional, it's not needed for many applications.

We will probably extend this to include things like the [scheme](https://github.com/gammapy/gammapy/blob/e287ca5d7df86189ee288ccefa8e2dbdbe4ac153/gammapy/obs/datastore.py#L128) and "chain", "version", "config" (see [here](https://hess-confluence.desy.de/confluence/display/HESS/HESS+Open+Source+Tools+-+HESS+data+and+IRFs+in+FITS+format#HESSOpenSourceTools-HESSdataandIRFsinFITSformat-File/folderformat)), i.e. all the info that's needed to access the right files via the `DataStore`.

@mapazarr - I know improving data format documentation and code in `gammapy.obs` is a lot of work, but this observation and data management stuff is actually the bigger part of building background models from off runs. At the core building background models just means stacking and smoothing histograms filled from event-lists, i.e. this will not be 1000s of lines of code and it's OK and even needed for you to spend more time on the observation and data management handling.

mapazarr replied:
Done.

cdeil commented on docs/dataformats/observation_lists.rst:
```diff
@@ -10,38 +10,87 @@ the information in the run list. These filenames can either be given directly
+
+Extra user defined header entries are allowed; Gammapy will ignore them.
+
+TODO: ``TEL_LIST``: since the final format is not clear, please mind having a look to this issue and its outcome:
```
If there's an issue, you can remove the TODO from here.
Github issues have milestones which make sure things aren't forgotten.
But forgetting to remove this TODO from this RST page once that issue is adressed is not unlikely.

I use TODOs (mostly in code comments) sometimes for things that I think are too small to make Github issues for. They can still be found and addressed systematically via full-text search. But it's probably not a great way to work and I should stop doing that.

mapazarr replied:
Done. It was your suggestion to put it on the first place :-)
It's true that TODO's are not very nice, when looking into the code, but I still think they're useful so that people are aware of what is missing. It can help in debugging a problem in the analysis sometimes.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +72,147 @@ def sigma_energy_theta(energy, theta, sigma):
+
+    # build a time reference as the start of 2010
+    dateref = Time('2010-01-01 00:00:00', format='iso', scale='utc')
+    # TODO: using format "iso" for `~astropy.Time` for now; eventually change it
```
I'm actually -1 on using the "fits" format, I think we should use "iso".
There's plenty of software (like python datetime or pandas) that can parse the iso format, the fits format is much less common.
The fact that we use UTC should be stated once in the docs and then it's OK.

mapazarr replied:
Done.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +72,147 @@ def sigma_energy_theta(energy, theta, sigma):
+    # in TIME_START and TIME_STOP
+    az = Angle(obs_table['AZ'])
+    alt = Angle(obs_table['ALT'])
+    if debug :
```
PEP8 will give a warning here, there should be no space before the colon:
`debug :` -> `debug:`

This is not very important, but please run `pep8` on this file to make sure your code formatting is OK.

mapazarr replied:
Done, done.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +72,147 @@ def sigma_energy_theta(energy, theta, sigma):
+    obs_table['RA'] = sky_coord.ra
+    obs_table['DEC'] = sky_coord.dec
+
+    # optional: it would be nice to plot a skymap with the simulated RA/dec positions
```
If you want to do this, do it in a script in `examples`.
But it will never be done at this point in the code, so please remove this comment.

mapazarr replied:
Removed.

cdeil commented on gammapy/datasets/tests/test_make.py:
```diff
@@ -15,3 +16,10 @@ def test_make_test_psf_fits_table():
+    observatory_name='HESS'
+    n_obs = 10
+    obs_table = make_test_observation_table(observatory_name, n_obs)
+    assert len(obs_table) == n_obs
```
Maybe add a few more asserts to check that positions and times are roughly as expected, i.e. that they were filled correctly?

mapazarr replied:
Done.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -11,7 +14,7 @@
 
 class Observation(object):
-    """Observation (a.k.a. Run).
+    """Observation (a.k.a. run).
```
If you look here: https://gammapy.readthedocs.org/en/latest/obs/index.html#classes
you'll see that in the short description of the class the line is truncated:
```
Observation (a.k.a.
```
This is because Sphinx or the Numpy docstring Sphinx extension looks for a dot `.` when parsing the docstring.

So the one-line description should always be ended with a dot, and can't have dots inside.
So please rephrase here, e.g. `"""A single observation."""` and then explain later that observation is called run by some people if you think it's useful.
This class might go away, so it's not super-important.

I'm just mentioning this here because I want you to become a Gammapy / Python / Sphinx pro and over the coming weeks help clean up and improve the code and docs where-ever you see an opportunity for improvement. We can only achieve a really nice package if we keep re-factoring and polishing and we need your help with that.

mapazarr replied:
Ok, fixed.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -38,24 +41,28 @@ class ObservationTable(Table):
-    * ...
-
+    convenience methods. The format of the observation table
+    is described in :ref:`dataformats_observation_lists`.
```
This is actually not obvious to me: does this link work in the HTML docs?
(I don't have time to build the docs locally and check myself.)

mapazarr replied:
Removed.

cdeil replied:
Yes, this is nice. One can even cross-link to Astropy or other doc sets via intersphinx.
I didn't know this a year ago, so there's plenty of places in Gammapy where one could cross-link high-level and API docs. Please do when you see opportunities for improvement or centralising duplicated documentation.

cdeil commented on gammapy/obs/observation.py:
```diff
@@ -38,24 +41,28 @@ class ObservationTable(Table):
+        time_ref = time_ref_from_dict(self.meta)
+        time_ref_unit = time_ref_from_dict(self.meta).format
+        ss += 'Time reference: {} {}'.format(time_ref, time_ref_unit)
+        # TODO: for an unknown reason, I can't enable astropy.units.cds
```
Let's try to reduce the number of new TODOs in Gammapy. The options are:
* You try to figure it out and start using it.
* Or, if it doesn't work, comment here what you've tried and want to achieve and I'll investigate.
* You let it go because we don't really need this at all and remove the TODO.

mapazarr replied:
Removed. I found a workaround.

cdeil commented on gammapy/utils/tests/test_time.py:
```diff
@@ -0,0 +1,27 @@
+    mjd_frac = 0.5
+    time_ref_dict = dict(MJDREFI=mjd_int, MJDREFF=mjd_frac)
+    time_ref = time_ref_from_dict(time_ref_dict)
+    decimal = 4
```
This temp variable doesn't help, you can save one line and just pass this as a keyword argument in the next line:
```python
assert_almost_equal(time_ref.mjd, mjd_int + mjd_frac, decimal=4)
```

mapazarr replied:
done

cdeil commented on gammapy/utils/tests/test_time.py:
```diff
@@ -0,0 +1,27 @@
+
+
+def test_time_relative_to_ref():
+    mjd_int = 500
```
Again, the lines where you set `mjd_int`, `mjd_frac` and `decimal` below can be removed, just pass the values directly as keyword arguments.

The main reasons to define temp variables are:
* You need the variable in many places in a piece of code
* Expressions become too long and need to be split up in pieces
* You pass values in somewhere in a function where you can only use positional arguments (i.e. have to use an un-pythonic, bad API) and not giving it a name reads to cryptic code (like you had for word / pix coordinate conversions).

mapazarr replied:
Ok.

cdeil commented:
@mapazarr – I did one more review of this PR and left inline comments. Nothing major, you should be able to finalise this PR today or tomorrow.
If you're not sure what exactly to write about times in Gammapy, just write two or three sentences ("should use Time class everywhere, when storing times in files use UTC or MET with a reference time") now and then we'll update this in the coming months when the relevant CTA spec is written by Catherine Boisson and me.

mapazarr commented:
OK, I implemented all issues. @cdeil: I hope you can now merge the PR.
I also figured out how to put the doc in google drive. It still is very slow in the upload, not because of the size of the docs dir, but because of the amount of files. @cdeil @adonath: do you know if there is a chance to upload compressed to google drive dirs and expand them once they are uploaded?
The link to the doc is here:
www.googledrive.com/host/0BzXHZQx0oCLBfmNwR294aFk5R0NuWlhTaDl5Q2Y0ajFJMGFXX0YxeW1CRlhjQ1FfelhFdEE/

cdeil commented on docs/dataformats/index.rst:
```diff
@@ -19,6 +19,16 @@ we simply use general-purpose file formats define our own semantics as described
+Time in Gammapy
+***************
+
+Gammapy adopts the exact same time standard as the Fermi Science Tools as described on this page: `Cicerone: Data - Time in Fermi Data Analysis <http://fermi.gsfc.nasa.gov/ssc/data/analysis/documentation/Cicerone/Cicerone_Data/Time_in_ScienceTools.html>`_.
```
Wrap lines to ~ 80 characters also in RST files.

This section looks out of place, first formats, then time, then you list the formats sub-pages TOC:
https://5c0a1090d7f5da950a4c3a6c72e5dfd9c101960f-www.googledrive.com/host/0BzXHZQx0oCLBfmNwR294aFk5R0NuWlhTaDl5Q2Y0ajFJMGFXX0YxeW1CRlhjQ1FfelhFdEE/dataformats/index.html#time-in-gammapy

mapazarr replied:
I thought you wanted to have it there?
 https://github.com/gammapy/gammapy/pull/278#discussion_r32211328
Please make a suggestion where should it go.

cdeil commented on gammapy/utils/time.py:
```diff
@@ -0,0 +1,58 @@
+
+    Returns
+    -------
+    `~astropy.time.TimeDelta` object with the time in seconds after the reference.
```
Use the normal numpy docstring format for this return value.

mapazarr replied:
Fixed.

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +72,146 @@ def sigma_energy_theta(energy, theta, sigma):
+    muon_efficiency_min = 0.6
+    muon_efficiency_max = 1.0
+    muon_efficiency = np.random.random(
+        len(obs_id)) * (muon_efficiency_max - muon_efficiency_min) + muon_efficiency_min
```
I don't like multi-line expressions. They are hard to read. With this line break especially.
Please use a temp variable instead to avoid the line break.

mapazarr replied:
I also don't, but I also don't like breaking a simple expression into many lines of code with temporary variables: it makes the code very hard to read. This was the suggestion of `autopep8` or long lines of code. I'd rather have it like this instead of with many intermediate steps.

cdeil replied:
Coding style and what is more or less readable code is always a matter of taste.
For now I'm calling the shots in Gammapy and am defining the coding style.
I'll clean this up in #285.

Note that this is not unusual, as developers join and leave to work on projects they have to adapt to the existing coding style or the one maintainers aim for even if they don't like it, because consistency is important.

:smile:

cdeil commented on gammapy/datasets/make.py:
```diff
@@ -64,3 +72,146 @@ def sigma_energy_theta(energy, theta, sigma):
+    obs_table['RA'] = sky_coord.ra
+    obs_table['DEC'] = sky_coord.dec
+
+    # positions
```
What's up with this comment? Remove?

mapazarr replied:
Fixed.

cdeil commented:
A few more very minor comments inline. I'll merge this tomorrow morning after travis-ci passes.

Yes, docs have lots of files and upload takes a few minutes. I think that's acceptable for large PRs where having the docs online helps.
As I said, I don't want to download the tarball ... if I have to get the docs locally I prefer checking out your branch via git and running `python setup.py build_sphinx`.

:sleeping:

cdeil commented:
One more thing ... I see you've edited the comment above.
Github doesn't send email notifications for comment edits, so usually I won't see any extra things you write. It's better to make a new comment if you forgot to say something so that I'll see it.

mapazarr commented:
I am aware of the error (warning) in the build_sphinx in travis-ci. I'm working on it.
One question:
I have a file called:
 docs/api/gammapy.datasets.generate_observation_table.rst
that issues a warning in my local build of sphinx. It comes from the time when `make_test_observation_table` was called `generate_observation_table`. But I didn't create the file myself. Can I just delete it? Or how do I clean all the references to the old name?

mapazarr commented:
@cdeil I implemented the comments and fixed the warning in the sphinx doc. Could you please have a look at my concern in my previous comment?
Here is the new doc. BTW it doesn't take minutes to upload, buth rather ~ 1 h:
 www.googledrive.com/host/0BzXHZQx0oCLBSGNKZDBZcnhlZzg/

cdeil commented:
@mapazarr - Thanks for all your work on this and your first big PR for Gammapy being merged.

* Forget about putting HTML docs online for code review then. I'll check them locally where needed.
* I realise now that time handling in Gammapy is too complicated to discuss and address it as a side project in this PR. So I've created #284 and #285 where I'll continue this.

Probably the same will happen with (RA, DEC) <-> (ALT/AZ) <-> (DETX/DETY) coordinates.
It's complicated in detail, even if not much code needs to be written.
It's also needed for background modeling, which should be done in the tangential system and then background models reprojected onto sky images.
Please think about whether you want to learn about coordinates and take on the implementation and documentation of the FOV <-> Sky coordinate transformations and reprojection, or if that is something I should do. We can discuss and decide this in the Google hangout next week.

cdeil commented:
Concerning your question about the old `docs/api/gammapy.datasets.generate_observation_table.rst` file, I've added an entry to the Gammapy developer docs here: https://github.com/gammapy/gammapy/pull/285/files#diff-00ab472a96028e9c3bcd77657d6c3701R15
